### PR TITLE
CI/BLD: use scipy-openblas wheel when building

### PR DESCRIPTION
**Pull request by mattip**
Replaces #20074. Uses the scipy-openblas32 wheel for building Scipy

The first commit changes linux wheel builds to use the scipy-openblas32 wheel

#### What does this implement/fix?

- ~Add a new cibw_before_build.sh script~ Change `cibw_before_build*.sh` scripts to use scipy-openblas wheels
- Remove the OpenBLAS version check from tests

#### Additional information

cc @rgommers

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -139,7 +138,7 @@ test-command = "bash {project}/tools/wheels/cibw_test_command.sh {project}"
 [tool.cibuildwheel.linux]
 manylinux-x86_64-image = "manylinux2014"
 manylinux-aarch64-image = "manylinux2014"
-before-build = "bash {project}/tools/wheels/cibw_before_build_linux.sh {project}"
+before-build = "bash {project}/tools/wheels/cibw_before_build.sh {project}"
 
 [tool.cibuildwheel.macos]
 before-build = "bash {project}/tools/wheels/cibw_before_build_macos.sh {project}"
```
Comment (andyfaff) [inline pyproject.toml:144]: This line will override any config specified in `[tool.cibuildwheel]` (i.e. the new file you add won't get called). Ditto the specific windows file.

For me it makes sense to separate configuration into separate `cibw_before_build_platform` files, a single file can get convoluted.

Comment (andyfaff) [reply pyproject.toml:144]: Perhaps there could be a common file, and individual files in this line, i.e.
```
before-build = "bash {project}/tools/wheels/cibw_before_build_common.sh; bash {project}/tools/wheels/cibw_before_build_macos.sh {project}"
```

Comment (mattip) [reply pyproject.toml:144]: > This line will override any config specified in `[tool.cibuildwheel]` (i.e. the new file you add won't get called).

At this time, I only want to override the linux wheel building to prove the wheels work. If it works, we can think about how to generalize it in another PR/other commits.

Comment (andyfaff) [reply pyproject.toml:144]: Ok. The changes you make here are also overridden in [wheels.yml](https://github.com/scipy/scipy/blob/85736e165b309ae57181c9c24f65c6a70b1eb7fa/.github/workflows/wheels.yml#L179)

Comment (mattip) [reply pyproject.toml:144]: Oy. I will try removing the redundant line from `wheels.yml`.

Comment (andyfaff) [reply pyproject.toml:144]: Having a duplicate is sometimes useful. e.g. if you want to develop locally you want things specified in pyproject.toml. If you want a fixed config for scipy/scipy to build wheels for public consumption you may want to do that in wheels.yml.

Comment (mattip) [reply pyproject.toml:151]: > For me it makes sense to separate configuration into separate ...

OK, adopted in the latest commit

Comment (mattip) [reply pyproject.toml:151]: > Having a duplicate is sometimes useful.

Hmm. On the other hand, it is a bit obscure and makes replicating the wheel build locally more difficult...

Comment (mattip) [reply pyproject.toml:151]: Maybe we could document the ability to override the pyproject.toml values with local environment variables, and local specialized builds would do that

Comment (andyfaff) [reply pyproject.toml:151]: Sure. The main reason to do so is e.g. employ specific package versions, etc.

**Comment by mattip** [comment]
The `scipy_` `BLAS_SYMBOL_PREFIX` is not getting prepended to [`srotg`](https://github.com/scipy/scipy/actions/runs/8843648177/job/24284175062#step:5:8561). I see that [scpiy-openblas 0.3.27](https://github.com/scipy/scipy/actions/runs/8843648177/job/24284175062#step:5:89) is found, which should result in using the cflags from [the pkgconfig file](https://github.com/MacPython/openblas-libs/blob/dd554116005d6431d182cc10f987304f358f70df/local/scipy_openblas64/__init__.py#L73). I wonder where that is getting lost.

**Review by rgommers — COMMENTED** [review]
This need for `LD_LIBRARY_PATH` is quite annoying. @ev-br pinged me about an issue with this as well yesterday. I haven't been able to reproduce it myself, but my conclusion was that `Libs: -L{libdir} -lscipy_openblas` in the generated `.pc` file is not reliable since it depends on the loader's library search path configuration as well as whether `site-packages` is under sysroot or not, and it should be replaced with `/abspath/to/scipy_openblas.so` to ensure the built extension modules will always find `scipy_openblas.so` at runtime.

All of the boilerplate of copying shared libraries around should be reduceable to:
```
$ python -c "import scipy_openblas32; print(scipy_openblas32.get_pkg_config())" > scipy-openblas.pc
$ export PKG_CONFIG_PATH=$PWD
```

Does that sound right to you @mattip?

**Comment by ev-br** [comment]
For LD_LIBRARY_PATH, my band-aid fix was to copy the  .so file from build/ to build-install:

https://github.com/ev-br/openblas-bench/blob/main/.github/workflows/codspeed.yml#L42

Still puzzling why the two libs have different linkage.

**Comment by mattip** [comment]
LD_LIBRARY_PATH is needed at runtime, not build time. The linker arguments in the pkg_config file are for build time, not runtime. In a proper wheel, delocate/auditwheel are used so LD_LIBRARY_PATH is no longer needed. But in the CI workflows, we need to either copy the dependencies into somwhere like /usr/local/lib, which is what the current scripts do, or use LD_LIBRARY_PATH

**Comment by rgommers** [comment]
I know it's a runtime thing, I am saying that it should not be needed. It is bad for a build to succeed and then fail at runtime.

> In a proper wheel, delocate/auditwheel are used 

We have a goal of wanting it to work without vendoring. Also, folks are trying to use the `scipy-openblas32`/`64` locally. And we want the cibuildwheel config to be as simple as possible. That's three reasons to make this work without `LD_LIBRARY_PATH`. 

I don't think it's hard. It should only require passing an absolute path rather than `-L`/`-l`, as I said above. We should be able to do this in `scipy_openblas32.get_pkg_config()`, right? It knows exactly where the shared library is located, so writing out the `.pc` file containing the correct absolute path seems like only a small tweak.

**Comment by mattip** [comment]
> writing out the .pc file containing the correct absolute path…

Makes sense. I can try this next week

**Comment by mattip** [comment]
> it should be replaced with /abspath/to/scipy_openblas.so to ensure the built extension modules will always find scipy_openblas.so at runtime

That seems to work on linux. How should this work on macos? Linking with the absolute path does not preserve the path for runtime, nor does using `-Wl,-rpath,/abspath/to` properly add an RPATH to the shared object (at least as far as I can tell).

**Comment by rgommers** [comment]
> nor does using `-Wl,-rpath,/abspath/`

This is what conda-forge compilers use on macOS, so I expect it to work:
```
% echo $LDFLAGS
-Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,/Users/rgommers/mambaforge/envs/scipy-dev/lib -L/Users/rgommers/mambaforge/envs/scipy-dev/lib
```
It would be surprising if it works when given via `LDFLAGS` (so for all targets) but not if it's only in the list of link args for the targets where it's needed. Are you sure it doesn't work? If so, does using `LDFLAGS` instead work?

**Comment by mattip** [comment]
The `-Wl,-rpath,<path>` part is needed in the `Libs:` line of `scipy-openblas.pc` to set the `rpath` search path. 
It doesn't matter whether I use the `full/path/to/libscipy_openblas.dylib` or `-L/path -lscipy_openblas`, the compiled linalg shared objects still only have a dependency on `libscipy_openblas.dylib`

The missing part is that I need to do this in order to "burn" the `@rpath` prefix onto the libscipy_openblas in the `LC_LOAD_DYLIB` stanza of the dylib header in all the shared objects that use scipy_openblas: 
```
install_name_tool -change libscipy_openblas.dylib @rpath/libscipy_openblas.dylib <path-to>/linalg/shared-object.cpython-311-darwin.so
```

I am still trying to figure out how to set the `@rpath` prefix from the linker without `install_name_tool`.

**Comment by rgommers** [comment]
The `-Wl,-rpath` addition is correct. The problem is with `libscipy_openblas.dylib`. Its install name (`LC_ID_DYLIB`) should include `@rpath`. You can check with:
```
% otool -D libscipy_openblas.dylib
libscipy_openblas.dylib:
libscipy_openblas.dylib
```
or with:
```
% otool -l libscipy_openblas64_.dylib | grep -A 2 LC_LOAD_DYLIB
```

To test that, you can fix it locally like this:
```
% install_name_tool -id @rpath/libscipy_openblas.dylib libscipy_openblas.dylib
% codesign --force -s - libscipy_openblas.dylib
```
After that, the install name is correct:
```
% otool -l libscipy_openblas.dylib | grep -A 2 LC_ID_DYLIB
          cmd LC_ID_DYLIB
      cmdsize 56
         name @rpath/libscipy_openblas.dylib (offset 24)
```
and adding `-Wl,-rpath,${libdir}` to the `.pc` file will work as expected. 

You can also try checking install names of for example the `<prefix>/lib` directories of conda-forge or Homebrew. For conda-forge, all dylibs start with `@rpath`. For Homebrew, they're all absolute paths. Both work - but a plain `libscipy_openblas.dylib` will not. The conda-forge approach is preferred (see, e.g., https://forums.developer.apple.com/forums/thread/736719 - generally a very useful article).

tl;dr rebuilding `scipy-openblas` with a change to the install name of the shared library on macOS, as well as including `-Wl,-rpath,${libdir}` for all platforms except Windows, should fix this.

**Comment by mattip** [comment]
> All of the boilerplate of copying shared libraries around should be reduceable to:
  ```
$ python -c "import scipy_openblas32; print(scipy_openblas32.get_pkg_config())" > scipy-openblas.pc
$ export PKG_CONFIG_PATH=$PWD
  ```

I updated the wheels so those two lines are all that is needed. The error which indicates something is not working right with the `-DBLAS_SYMBOL_PREFIX=scipy_` is still there:
```
ImportError: Error relocating /__w/scipy/scipy/build-install/lib/python3.10/site-packages/scipy/linalg/_fblas.cpython-310-x86_64-linux-gnu.so: srotg_: symbol not found
```

File: pyproject.toml (hunk 1 above)
Comment (rgommers) [reply pyproject.toml:151]: I like @mattip's idea. The duplication makes the code harder to understand, and I'd really like it to be as clean as possible. So comments only to "how do I tweak this if needed" are nicer than more bash code.

File: pyproject.toml [hunk 2]
```diff
@@ -141,6 +140,13 @@ manylinux-x86_64-image = "manylinux2014"
 manylinux-aarch64-image = "manylinux2014"
 before-build = "bash {project}/tools/wheels/cibw_before_build_linux.sh {project}"
 
+[tool.cibuildwheel.linux.environment]
+# RUNNER_OS is a GitHub Actions specific env var; define it here so it works on Cirrus CI too
+RUNNER_OS="Linux"
+# /project will be the $PWD equivalent inside the docker used to build the wheel
+PKG_CONFIG_PATH="/project/.openblas"
+LD_LIBRARY_PATH="$LD_LIBRARY_PATH:/project/.openblas/lib"
```
Comment (rgommers) [inline pyproject.toml:148]: This can go now, right? The `PKG_CONFIG_PATH` is maybe still needed (?), but the rest of all this should not be.

File: tools/wheels/cibw_before_build_linux.sh [hunk 3]
```diff
@@ -13,13 +13,13 @@ else
...
 cat $PROJECT_DIR/tools/wheels/LICENSE_linux.txt >> $PROJECT_DIR/LICENSE.txt
 
 # Install Openblas
-basedir=$(python tools/openblas_support.py $NIGHTLY_FLAG)
-cp -r $basedir/lib/* /usr/local/lib
-cp $basedir/include/* /usr/local/include
+echo PKG_CONFIG_PATH $PKG_CONFIG_PATH
+PKG_CONFIG_PATH=$PROJECT_DIR/.openblas
+rm -rf $PKG_CONFIG_PATH
+mkdir -p $PKG_CONFIG_PATH
```
Comment (rgommers) [inline tools/wheels/cibw_before_build_linux.sh:24]: The `rm -rf` and `mkdir` at least should no longer be needed. Also maybe change `PKG_CONFIG_PATH` to simply `$PWD`, so the separate `.openblas` directory isn't needed? A single `.pc` file can live anywhere, no need for a subdir.

**Comment by rgommers** [comment]
Great, much nicer! I'll try to have a look at the `srotg` issue tonight or tomorrow. It doesn't look like we treat it in a special way, however a comment in `npy_blas_base.h` says: `Routines with S and D prefix only`.

The macOS job failed, detection didn't work:
```
scipy/meson.build:191:9: ERROR: Dependency "OpenBLAS" not found, tried pkgconfig, framework and cmake
```
but I assume that's because this is now a pared-down PR focused on Linux?

**Comment by mattip** [comment]
Yes, once the speediest build job to use a scipy-openblas wheel works (which I think happens to be the musllinux one), I will clean this all up for the other build jobs and wheel builds.

**Comment by mattip** [comment]
If it helps to debug the problem, I can clean up the macos one so it too fails to apply the prefix?

**Comment by thalassemia** [comment]
The reason the symbol prefix is not being honored is because we currently only generate the required BLAS/LAPACK wrappers for Accelerate builds to handle their special symbol suffixes. Since scipy-openblas has a special symbol prefix, we can fix this by enabling wrapper generation for those builds as well (the commit I added). The wrappers should only be generated when there is a symbol prefix or suffix because they result in infinite recursion when the wrapper symbols are the same as the BLAS/LAPACK symbols (as they are when there is no prefix/suffix).

Could we automatically detect whether a BLAS/LAPACK library has prefixes/suffixes instead of enabling wrapper generation on a case-by-case basis?

**Comment by rgommers** [comment]
Ah of course, good catch @thalassemia.

> Could we automatically detect whether a BLAS/LAPACK library has prefixes/suffixes instead of enabling wrapper generation on a case-by-case basis?

Not fully automatically, but we can specify a set of "these are the set of potential prefixes/suffixes to check". E.g.: https://github.com/numpy/meson/blob/4e370ca8ab73c07f7b84abe8a4b937caace050a4/mesonbuild/dependencies/blas_lapack.py#L398-L409

For now doing it manually is the way to go, but we can add a standard set to check in Meson. I think it mainly matters for OpenBLAS.

**Comment by rgommers** [comment]
The `musllinux` job is happy now. I cancelled the two _Prerelease deps_ jobs, because they were hanging - should be related, since they were using an older version of `scipy-openblas`, so the prefix change probably created a conflict there. Anyway, this should be unblocked now to update the other jobs.

File: pyproject.toml (hunk 2 above)
Comment (mattip) [reply pyproject.toml:148]: fixed

File: tools/wheels/cibw_before_build_linux.sh (hunk 3 above)
Comment (mattip) [reply tools/wheels/cibw_before_build_linux.sh:24]: adopted

**Comment by mattip** [comment]
Locally running the 32-bit linux in a docker passes tests. I wonder what went wrong in the CI.

**Comment by mattip** [comment]
The 32-bit tests get stuck in `scipy/cluster/tests/test_vq.py::TestKMean::test_kmeans2_high_dim[numpy]`

**Comment by thalassemia** [comment]
I can reproduce the 32-bit stall in a local Docker image. Stepping through with a debugger, execution stalls after this line: https://github.com/scipy/scipy/blob/0f8796680a5e5572c71b5f56f30a46da1901b85e/scipy/cluster/vq.py#L561
Stepping into the Numpy code, I identified `numpy.linalg._umath_linalg.svd_m_s` [as the stalled function](https://github.com/numpy/numpy/blob/de82cd9468704a033702974010ee7e7efc85b393/numpy/linalg/linalg.py#L1652-L1657). 

When I get some time later, I'll try bumping the Numpy version and/or installing `scipy-openblas` before Numpy to see if either helps.

**Comment by mattip** [comment]
Two of the CI failures I think are not related:
- "Windows tests / full, py3.10/npMin, dev.py" (timeouts in tests) is being investigated in #20699
- "macOS tests / Conda & umfpack/scikit-sparse, fast, py3.11/npAny, dev.py"  is failing to resolve `conda create` everywhere

"Linux tests / 32-bit, fast, py3.10/npMin, dev.py" hangs somewhere in numpy, which is being built from source on every PR run since NumPy does not release wheels for 32-bit linux. Before this PR, it would pick up the openblas installed from the openblas_support script. Now the build is not using any accelerator, which may be the reason for the hang.

**Comment by mattip** [comment]
Maybe 32bit SciPy is incompatible with NumPy2.0.0? This line https://github.com/scipy/scipy/blob/7febfdd1c9b477dcf89b103d421afb44c7e822fd/scipy/stats/_stats_py.py#L1085 is getting executed with `dtype=int32` which fails (`finfo(int)` raises an error). Earlier there is this https://github.com/scipy/scipy/blob/7febfdd1c9b477dcf89b103d421afb44c7e822fd/scipy/stats/_stats_py.py#L1050-L1054. That goes into NumPy, to where it checks dtype against NumPy's `sctypes["int"]`, which is `[<class 'numpy.int8'>, <class 'numpy.int16'>, <class 'numpy.intc'>, <class 'numpy.int64'>]`. Note `int32` is missing. I will raise an issue on NumPy.

**Comment by thalassemia** [comment]
I got Numpy 1.23.5 to build with `scipy-openblas` by writing a custom `$HOME/.numpy-site.cfg` (9c28277). It passes on my copy of this branch. I did try bumping the version to 2.0.0rc2 as @mittip did but saw test failures/crashing, so I decided to pursue this route instead.

**Comment by mattip** [comment]
> writing a custom `$HOME/.numpy-site.cfg`

Brilliant, thanks. Indeed, "Linux tests / 32-bit" is now passing. The remaining two CI failures are not related to this PR. However there is #20718 about build times being slower when using the `dev.py --with-scipy-openblas` workflow.

**Comment by rgommers** [comment]
The 45 failures on Windows that are timing-related do look like an issue with this PR (or this `scipy-openblas32` build). Here are two examples:
```
 _________________________ TestLstsq.test_random_exact _________________________
[gw1] win32 -- Python 3.10.11 C:\hostedtoolcache\windows\Python\3.10.11\x64\python.exe
Test passed but took too long to run: Duration 27.418427899999642s > 5.0s

_________________________ TestSVDVals.test_crash_2609 _________________________
[gw3] win32 -- Python 3.10.11 C:\hostedtoolcache\windows\Python\3.10.11\x64\python.exe
Test passed but took too long to run: Duration 262.60931979999987s > 5.0s
```

Comparing with [a recent run on `main`](https://github.com/scipy/scipy/actions/runs/9107388442/job/25036195235) shows they're now 10x - 100x slower. That hints at a lock contention or symbol clash issue between the two `scipy-openblas` builds included in `numpy` and `scipy`:
```
 2.03s call     build-install/Lib/site-packages/scipy/linalg/tests/test_basic.py::TestLstsq::test_random_exact
 1.77s call     build-install/Lib/site-packages/scipy/linalg/tests/test_decomp.py::TestSVDVals::test_crash_2609
```

**Comment by mattip** [comment]
rebased off main to get the fix in #20699

**Comment by mattip** [comment]
That solved the slow tests. Now a [test is failing](https://github.com/scipy/scipy/actions/runs/9207341724/job/25327176234?pr=20585#step:8:1261) on numpy/numpy#26439:
```
ValueError: data type <class 'numpy.int32'> not inexact
```

**Comment by rgommers** [comment]
> Should this PR remove the openblas preference from the meson defaults? It seems to be working when using scipy-openblas regardless of this setting

I don't think so. A PR for auto-detection in the same way as NumPy does it is still to be opened. That will overhaul all this, for now I don't see a reason to touch it.

**Comment by mattip** [comment]
Ci (except for "owner approval") is passing. Should I trigger the wheel building, or leave it for a future PR?

**Comment by rgommers** [comment]
Great! Please do trigger the wheel builds, since the changes here affect wheel builds.

**Comment by mattip** [comment]
CI is passing, I think this might be ready for a review.

**Comment by mattip** [comment]
There was a merge conflict, so I rebased, cleaned this up a bit, and squashed all the commits.

File: .github/workflows/windows.yml [hunk 4]
```diff
@@ -49,14 +49,16 @@ jobs:
 
       - name: pip-packages
         run: |
-          pip install numpy cython pybind11 pythran meson ninja pytest pytest-xdist pytest-timeout pytest-fail-slow pooch rich_click click doit pydevtool hypothesis "scipy-openblas32<=0.3.23.293.2"
+          pip install numpy cython pybind11 pythran meson ninja pytest pytest-xdist pytest-timeout pytest-fail-slow pooch rich_click click doit pydevtool hypothesis
```
Comment (rgommers) [inline .github/workflows/windows.yml:52]: This should have kept `scipy-openblas32`

Comment (rgommers) [reply .github/workflows/windows.yml:52]: Or `python -m pip install -r requirements/openblas.txt` like in the next job

File: .github/workflows/windows.yml [hunk 5]
```diff
@@ -89,14 +91,17 @@ jobs:
...
+          python -m pip install numpy==1.23.5 cython pybind11 pythran meson-python meson ninja pytest pytest-xdist pytest-timeout pytest-fail-slow pooch rich_click click doit pydevtool hypothesis
+          python -m pip install -r requirements/openblas.txt
 
       - name: Build
         run: |
           python dev.py build --with-scipy-openblas
 
       - name: Test
         run: |
+          # test runner parallel clashes with OpenBLAS multithreding
```
Comment (rgommers) [inline .github/workflows/windows.yml:103]: nit: typo in "multithreading"

File: .github/workflows/windows.yml [hunk 6]
```diff
@@ -130,30 +135,34 @@ jobs:
...
+          delvewheel repair --add-path $openblas_dir --no-dll libsf_error_state.dll -w wheelhouse $wheel_name
           
-          python -m pip install $env:wheel_name
+          python -m pip install wheelhouse/*
 
       - name: Test
         run: |
           cd $RUNNER_TEMP
           # run full test suite
+          # test runner parallel clashes with OpenBLAS multithreding
```
Comment (rgommers) [inline .github/workflows/windows.yml:166]: same typo here

File: pyproject.toml [hunk 7]
```diff
@@ -141,22 +140,24 @@ manylinux-x86_64-image = "manylinux2014"
...
+# /project will be the $PWD equivalent inside the docker used to build the wheel
+PKG_CONFIG_PATH="/project/"
+
 [tool.cibuildwheel.macos]
-before-build = "bash {project}/tools/wheels/cibw_before_build_macos.sh {project}"
+before-build = "pip install ninja; bash {project}/tools/wheels/cibw_before_build_macos.sh {project}"
+
+[tool.cibuildwheel.macos.environment]
+# On macos-arm64, use Accelerate instead of setting this
+# CIBW_CONFIG_SETTINGS=\"setup-args=-Dblas=accelerate
```
Comment (rgommers) [inline pyproject.toml:152]: What is happening with this comment?

File: pyproject.toml [hunk 8]
```diff
@@ -141,22 +140,24 @@ manylinux-x86_64-image = "manylinux2014"
...
+PKG_CONFIG_PATH="{project}"
 
 [tool.cibuildwheel.windows]
-before-build = "bash {project}/tools/wheels/cibw_before_build_win.sh {project}"
+before-build = "pip install ninja && bash {project}/tools/wheels/cibw_before_build_win.sh {project}"
```
Comment (rgommers) [inline pyproject.toml:156]: This change should not be needed. Probably was a temporary change when you had build isolation disabled?

File: pyproject.toml [hunk 9]
```diff
@@ -141,22 +140,24 @@ manylinux-x86_64-image = "manylinux2014"
...
-# An alternative is to set CMAKE_PREFIX_PATH="c:/opt/openblas/if_32/32"
-# Don't use double backslash for path separators, they don't get passed
-# to the build correctly
-# environment = { CMAKE_PREFIX_PATH="c:/opt/64" }
-environment = { PKG_CONFIG_PATH = "c:/opt/64/lib/pkgconfig" }
+[tool.cibuildwheel.windows.environment]
+# This does not work because pkg-config does not like backslashes,
+PKG_CONFIG_PATH="{project}"
+# do this instead (which will override this setting)
+# set CIBW_ENVIRONMENT_WINDOWS=PKG_CONFIG_PATH=PWD.replace('\\', '/')
```
Comment (rgommers) [inline pyproject.toml:163]: what is happening with this comment?

File: tools/wheels/cibw_test_command.sh [hunk 10]
```diff
@@ -3,10 +3,8 @@ set -xe
 PROJECT_DIR="$1"
 
 # python $PROJECT_DIR/tools/wheels/check_license.py
-if [[ $(uname) == "Linux" ]] ; then
-    python $PROJECT_DIR/tools/openblas_support.py --check_version
-fi
-echo $?
+# if [[ $(uname) == "Linux" ]] ; then
+    # python $PROJECT_DIR/tools/openblas_support.py --check_version
+# fi
```
Comment (rgommers) [inline tools/wheels/cibw_test_command.sh:8]: delete?

**Comment by rgommers** [comment]
Thanks Matti! This looks quite good, and makes things significantly easier to understand.

File: pyproject.toml (hunk 7 above)
Comment (mattip) [reply pyproject.toml:152]: Deleted, it was left over from some previous iteration of wheel build settings

File: pyproject.toml (hunk 8 above)
Comment (mattip) [reply pyproject.toml:156]: Let's see what happens if I remove it. Does anything else install ninja on windows and macos?

File: pyproject.toml (hunk 9 above)
Comment (mattip) [reply pyproject.toml:163]: You cannot set `PKG_CONFIG_PATH` to `"{project}"` on windows, the path native separators confuse `pkg-config`. So you have to set it via CIBW_ENVIRONMENT_WINDOWS instead. This is a comment for people wishing to build wheels that they must use that alternative. Should I remove it?

File: pyproject.toml (hunk 8 above)
Comment (rgommers) [reply pyproject.toml]: Yes, `meson-python` will unless build isolation is disabled. CI seems happy, so resolving.

File: pyproject.toml (hunk 9 above)
Comment (rgommers) [reply pyproject.toml:161]: Ah okay. I didn't really understand that from the comment - let's just clarify it instead.

To be honest I also don't understand the "this does not work" comment (which I now think is also aimed at folks wanting to build locally?), nor what `"{project}"` expands to and which tool does the expanding.

Comment (rgommers) [reply pyproject.toml:161]: Okay, let's deal with this loose end post-merge. Time to get this in, so we have it in the 1.14.x branch - that'll make maintenance easier.

**Review by rgommers — APPROVED** [review]
Nice work, thanks for getting this over the finish line @mattip! And thanks for the reviews @andyfaff

**Comment by mattip** [comment]
It seems some of the windows tests are still slower now, there are occasional CI failures due to test durations being over the limits. Was that happening before this PR?

**Comment by rgommers** [comment]
I think what you're seeing is gh-20806, and that's due to CI machinery changes unrelated to this PR.
